toolchain/check/deduce.cpp

```diff
@@ -75,6 +75,60 @@ class DeductionWorklist {
   Context& context_;
   llvm::SmallVector<PendingDeduction> deductions_;
 };
+
+// State that is tracked throughout the deduction process.
+class DeductionContext {
+ public:
+  DeductionContext(Context& context, SemIR::LocId loc_id,
+                   SemIR::GenericId generic_id, bool diagnose)
+      : context_(&context),
+        loc_id_(loc_id),
+        generic_id_(generic_id),
+        diagnose_(diagnose) {
+    CARBON_CHECK(generic_id.is_valid(),
+                 "Performing deduction for non-generic entity");
+
+    // Initialize the deduced arguments to Invalid.
+    result_arg_ids_.resize(
+        context.inst_blocks()
+            .Get(context.generics().Get(generic_id_).bindings_id)
+            .size(),
+        SemIR::InstId::Invalid);
+  }
+
+  auto context() const -> Context& { return *context_; }
+  auto worklist() -> DeductionWorklist& { return worklist_; }
+
+  // Adds the arguments from the given specific as known enclosing generic
+  // arguments that will not be deduced.
+  auto AddEnclosingArgs(SemIR::SpecificId enclosing_specific_id) -> void;
```

Comment:
Sure, done. I'd arranged it this way because `impl`s don't have an enclosing specific, but I suppose it's still meaningful to ask for one in all cases.